bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java

```diff
@@ -0,0 +1,182 @@
+        }
+
+        for (final Channel c : getThing().getChannels()) {
+            final ChannelTypeUID type = c.getChannelTypeUID();
+            if (type != null) {
+                final ChannelConfig channelConfig = c.getConfiguration().as(ChannelConfig.class);
+                try {
+                    final DeviceChannel deviceChannel = DeviceChannelFactory
+                            .createDeviceChannel(valueTransformationProvider, channelConfig, type.getId());
+                    if (deviceChannel != null) {
+                        channels.put(c.getUID(), deviceChannel);
+                    }
+                } catch (final IllegalArgumentException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Configuration error for channel " + c.getUID().getId() + ": " + e.getMessage());
+                    return;
+                }
+            }
+        }
+
+        updateStatus(ThingStatus.UNKNOWN);
+        bridgeStatusChanged(getBridgeStatus());
+    }
+
+    @Override
+    public void dispose() {
+        channels.clear();
+        data = null;
+        super.dispose();
+    }
+
+    /**
+     * Handle a line of data received from the bridge
+     *
+     * @param data the line of data
+     */
+    public void handleData(final String data) {
+        if (devicePattern.matcher(data).matches()) {
+            this.data = data;
+            channels.keySet().forEach(this::refresh);
+        }
+    }
+
+    @Override
+    public void bridgeStatusChanged(final ThingStatusInfo bridgeStatusInfo) {
+        if (getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.CONFIGURATION_ERROR) {
+            return;
+        }
+
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE && getThing().getStatus() == ThingStatus.UNKNOWN) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+            return;
+        }
+
+        super.bridgeStatusChanged(bridgeStatusInfo);
+    }
+
+    /**
+     * Return the bridge status.
+     */
+    private ThingStatusInfo getBridgeStatus() {
+        final Bridge b = getBridge();
+        if (b != null) {
+            return b.getStatusInfo();
+        } else {
+            return new ThingStatusInfo(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, null);
+        }
+    }
+
+    /**
+     * Refreshes the channel with the last received data
+     *
+     * @param channelId the channel to refresh
+     */
+    private void refresh(final ChannelUID channelUID) {
+        final String data = this.data;
+
+        if (data == null || !isLinked(channelUID)) {
+            return;
+        }
+
+        final DeviceChannel channel = channels.get(channelUID);
+        if (channel != null) {
+            final String state = channel.transformData(data);
+
+            if (state != null) {
+                updateState(channelUID, new StringType(state));
+            }
+        }
```

Comment:
```suggestion
        if (channel != null) {
            channel.transformData(data).ifPresent(value -> updateState(channelUID, new StringType(value));
        }
```

You have a null-check in l.169, so requiring non-null for the parameter should be ok. I think the same applies to `mapCommand`, please check that.